controllers/metal3.io/hostfirmwaresettings_controller.go

```diff
@@ -0,0 +1,295 @@
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	metal3v1alpha1 "github.com/metal3-io/baremetal-operator/apis/metal3.io/v1alpha1"
+	"github.com/metal3-io/baremetal-operator/pkg/provisioner"
+)
+
+// HostFirmwareSettingsReconciler reconciles a HostFirmwareSettings object
+type HostFirmwareSettingsReconciler struct {
+	client.Client
+	Log                logr.Logger
+	ProvisionerFactory provisioner.Factory
+}
+
+type rInfo struct {
+	log logr.Logger
+	hfs *metal3v1alpha1.HostFirmwareSettings
+}
+
+//+kubebuilder:rbac:groups=metal3.io,resources=hostfirmwaresettings,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=metal3.io,resources=hostfirmwaresettings/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=metal3.io,resources=firmwareschemas,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=metal3.io,resources=firmwareschemas/status,verbs=get;update;patch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.6.4/pkg/reconcile
+func (r *HostFirmwareSettingsReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, err error) {
+
+	reqLogger := r.Log.WithValues("hostfirmwaresettings", req.NamespacedName)
+	reqLogger.Info("start")
+
+	// Fetch the HostFirmwareSettings
+	hfs := &metal3v1alpha1.HostFirmwareSettings{}
+	if err = r.Get(ctx, req.NamespacedName, hfs); err != nil {
+		if k8serrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after
+			// reconcile request. Return and don't requeue
+			reqLogger.Info("could not find host firmware settings")
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, errors.Wrap(err, "could not load host firmware settings")
+	}
+
+	// Get settings from provisioner if update requested or settings are empty
+	if hfs.Status.ProvStatus.Update == true || hfs.Status.Settings == nil {
+
+		// Use the provisioner ID stored in annotation to access Ironic API
+		if _, ok := hfs.Annotations[metal3v1alpha1.ProvisionerIdAnnotation]; ok {
+			provID := hfs.Annotations[metal3v1alpha1.ProvisionerIdAnnotation]
+
+			info := &rInfo{log: reqLogger, hfs: hfs}
+
+			prov, err := r.ProvisionerFactory.NewProvisioner(provisioner.BuildEmptyHostData(provID), func(reason, message string) {})
```

Comment:
Good idea! For safey, it would be recommended to invoked the `IsReady()` method before using it

Reply:
Added the check for IsReady() along with a reschedule if not ready.